File: realtime-whisper-aurix/apps/frontend/electron/vector-service.ts
import { QdrantClient } from '@qdrant/js-client-rest';

const DOCUMENTATION_COLLECTION = 'documentation_collection';
const CONVERSATION_COLLECTION = 'conversation_collection';
const VECTOR_SIZE = 1024;

interface VectorPayload {
  [key: string]: unknown;
}

class VectorService {
  private client: QdrantClient;

  constructor(url: string, apiKey?: string) {
    this.client = new QdrantClient({
      url,
      ...(apiKey ? { apiKey } : {}),
    });
  }

  async initialize(): Promise<void> {
    await this.ensureCollection(DOCUMENTATION_COLLECTION, ['content_type', 'source', 'timestamp']);
    await this.ensureCollection(CONVERSATION_COLLECTION, [
      'content_type',
      'user_id',
      'conversation_id',
      'message_id',
      'role',
      'timestamp',
    ]);
    console.log('Vector service initialized');
  }

  private async ensureCollection(name: string, indexFields: string[]): Promise<void> {
    try {
      await this.client.getCollection(name);
    } catch {
      await this.client.createCollection(name, {
        vectors: { size: VECTOR_SIZE, distance: 'Cosine' },
      });
      for (const field of indexFields) {
        await this.client.createPayloadIndex(name, {
          field_name: field,
          field_schema: 'keyword',
        });
      }
    }
  }

  async upsertVector(
    collection: string,
    id: string,
    vector: number[],
    payload: VectorPayload,
  ): Promise<void> {
    await this.client.upsert(collection, {
      points: [{ id, vector, payload }],
    });
  }

  async upsertVectors(
    collection: string,
    points: { id: string; vector: number[]; payload: VectorPayload }[],
  ): Promise<void> {
    if (points.length === 0) return;
    await this.client.upsert(collection, { points });
  }

  async search(
    collection: string,
    vector: number[],
    limit: number = 5,
    filter?: Record<string, unknown>,
  ): Promise<Array<{ id: string | number; score: number; payload: VectorPayload }>> {
    const result = await this.client.search(collection, {
      vector,
      limit,
      with_payload: true,
      ...(filter ? { filter } : {}),
    });

    return result.map((r) => ({
      id: r.id,
      score: r.score,
      payload: (r.payload as VectorPayload) || {},
    }));
  }

  async deleteVectors(collection: string, ids: string[]): Promise<void> {
    if (ids.length === 0) return;
    await this.client.delete(collection, {
      points: ids,
    });
  }

  async getCollSectionInfo(collection: string) {
    return this.client.getCollection(collection);
  }
}

let vectorService: VectorService | null = null;

export function initializeVectorService(url: string, apiKey?: string): VectorService {
  vectorService = new VectorService(url, apiKey);
  return vectorService;
}

export function getVectorService(): VectorService | null {
  return vectorService;
}

export async function initializeVectorCollections(): Promise<void> {
  if (!vectorService) throw new Error('Vector service not initialized');
  await vectorService.initialize();
}

export { DOCUMENTATION_COLLECTION, CONVERSATION_COLLECTION };

File: realtime-whisper-aurix/apps/backend/app/services/vector_service.py
"""
Qdrant Vector Database Service
"""

from qdrant_client import QdrantClient
from qdrant_client.models import Distance, VectorParams, PointStruct
from app.core.config import settings
from typing import List, Dict, Any
import uuid


class VectorService:
    """Service for Qdrant vector database operations"""

    def __init__(self):
        self.client = QdrantClient(
            url=settings.QDRANT_URL,
            api_key=settings.QDRANT_API_KEY if settings.QDRANT_API_KEY else None
        )
        self.collection_name = settings.QDRANT_COLLECTION

    async def ensure_collection(self, vector_size: int = 1024):
        """
        Ensure collection exists, create if not

        Args:
            vector_size: Dimension of embedding vectors (Cohere v3.0 = 1024)
        """
        try:
            collections = self.client.get_collections().collections
            collection_names = [c.name for c in collections]

            if self.collection_name not in collection_names:
                self.client.create_collection(
                    collection_name=self.collection_name,
                    vectors_config=VectorParams(
                        size=vector_size,
                        distance=Distance.COSINE
                    )
                )

        except Exception as e:
            raise Exception(f"Collection creation failed: {str(e)}")

    async def insert_vectors(
        self,
        vectors: List[List[float]],
        payloads: List[Dict[str, Any]]
    ) -> List[str]:
        """
        Insert vectors into collection

        Args:
            vectors: List of embedding vectors
            payloads: List of metadata for each vector

        Returns:
            List of inserted point IDs
        """
        try:
            await self.ensure_collection()

            points = []
            ids = []

            for vector, payload in zip(vectors, payloads):
                point_id = str(uuid.uuid4())
                ids.append(point_id)
                points.append(
                    PointStruct(
                        id=point_id,
                        vector=vector,
                        payload=payload
                    )
                )

            self.client.upsert(
                collection_name=self.collection_name,
                points=points
            )

            return ids

        except Exception as e:
            raise Exception(f"Vector insertion failed: {str(e)}")

    async def search_vectors(
        self,
        query_vector: List[float],
        limit: int = 5,
        score_threshold: float = 0.7
    ) -> List[Dict[str, Any]]:
        """
        Search for similar vectors

        Args:
            query_vector: Query embedding vector
            limit: Maximum number of results
            score_threshold: Minimum similarity score

        Returns:
            List of search results with scores and payloads
        """
        try:
            results = self.client.search(
                collection_name=self.collection_name,
                query_vector=query_vector,
                limit=limit,
                score_threshold=score_threshold
            )

            return [
                {
                    "id": str(result.id),
                    "score": result.score,
                    "payload": result.payload
                }
                for result in results
            ]

        except Exception as e:
            raise Exception(f"Vector search failed: {str(e)}")

    async def get_collection_info(self) -> Dict[str, Any]:
        """Get information about the collection"""
        try:
            info = self.client.get_collection(self.collection_name)
            return {
                "name": self.collection_name,
                "vectors_count": info.vectors_count,
                "points_count": info.points_count,
                "status": info.status
            }
        except Exception as e:
            return {"error": str(e)}
